File: prototype/rails/src/app/domain/auth/magic_link_token.rb
require "digest"

module Domain
  module Auth
    # Only the digest is stored, so a leaked database row cannot be replayed as
    # a link.
    module MagicLinkToken
      module_function

      def digest(token)
        Digest::SHA256.hexdigest(token.to_s)
      end
    end
  end
end

File: prototype/php/src/app/Domain/Auth/MagicLinkToken.php
<?php

declare(strict_types=1);

namespace App\Domain\Auth;

final class MagicLinkToken
{
    private function __construct() {} // @codeCoverageIgnore

    /**
     * Only the digest is stored, so a leaked database row cannot be replayed
     * as a link.
     */
    public static function hash(string $token): string
    {
        return hash('sha256', $token);
    }
}
